Ability to delete multiple jobs by passing multiple ids

AleksandrMikhin opened the pull request:
This PR brings an ability to delete jobs by passing multi IDs.

Short description:
- implemented an ability to delite jobs by passing multi IDs
- changed the `DeletedJobInfo` class and the job deletion response
- fixed obtaining an owner of a job when it's deleted [SGE]

additionally:
- made tests refactoring of class `SgeJobProviderTest` class
- improved validation of ParallelEnvOptions passed to job submitting requests [SGE]
- added gradle task to run single tests from IDE without filtration

Commit:
moved the obtaining of job logs to a separate provider class

Commit:
Merge branch 'slurm_develop' into #10_slurm_ability_to_get_logs_for_specific_job

# Conflicts:
#	src/main/java/com/epam/grid/engine/service/JobOperationProviderService.java
#	src/main/resources/templates/slurm/command/sbatch

Commit:
apply ability to use a relative log path to the jobLog provider, a few improvements of build.gradle

Commit:
Merge branch 'slurm_develop' into #10_slurm_ability_to_get_logs_for_specific_job

# Conflicts:
#	src/main/java/com/epam/grid/engine/provider/job/slurm/SlurmJobProvider.java

Commit:
minor changes

Commit:
Merge branch 'slurm_develop' into #10_slurm_ability_to_get_logs_for_specific_job

Commit:
changed jobId to long into the JobLogProvider

Commit:
added a new feature to delete jobs by passing multi ids, made some tests refactoring

SilinPavel commented on src/main/java/com/epam/grid/engine/provider/job/sge/SgeJobProvider.java:
```diff
@@ -161,19 +159,48 @@ public CommandType getProviderType() {
+        if (result.getExitCode() != 0) {
+            CommandsUtils.throwExecutionDetails(result, HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+        final Matcher matcher = SUBMITTED_JOB_ID_PATTERN.matcher(result.getStdOut().get(0));
```
possible NPE here, when `get(0)` is called

AleksandrMikhin replied:
fixed the if-condition

SilinPavel commented on src/main/java/com/epam/grid/engine/provider/job/sge/SgeJobProvider.java:
```diff
@@ -236,85 +255,50 @@ private String[] makeQdelCommand(final DeleteJobFilter filter) {
      */
     private void validateDeleteRequest(final DeleteJobFilter deleteJobFilter) {
-        if (!StringUtils.hasText(deleteJobFilter.getUser()) && deleteJobFilter.getId() == null) {
+        if (!StringUtils.hasText(deleteJobFilter.getUser()) && CollectionUtils.isEmpty(deleteJobFilter.getIds())) {
```
how system will behave if we pass `owner` as well as `ids`, but ids weren't launched from `owner`? I believe it is better to check it here to, and do not allow run request if both things are specified

SilinPavel replied:
I also believe that such check could be moved to service layer to reuse this code

AleksandrMikhin replied:
done

Commit:
changes according to review comments

Commit:
fixed some tests,

merge branch 'slurm_develop' into ability_to_delete_multiple_jobs_by_passing_multiple_ids

# Conflicts:
#	src/main/java/com/epam/grid/engine/provider/job/JobProvider.java
#	src/main/java/com/epam/grid/engine/provider/job/sge/SgeJobProvider.java
#	src/main/java/com/epam/grid/engine/provider/job/slurm/SlurmJobProvider.java
#	src/main/java/com/epam/grid/engine/provider/log/JobLogProvider.java
#	src/main/java/com/epam/grid/engine/provider/log/JobLogProviderImpl.java
#	src/main/java/com/epam/grid/engine/service/JobOperationProviderService.java
#	src/main/resources/templates/slurm/command/sbatch
#	src/test/java/com/epam/grid/engine/provider/job/sge/SgeJobProviderTest.java
#	src/test/java/com/epam/grid/engine/provider/job/slurm/SlurmJobProviderTest.java
#	src/test/java/com/epam/grid/engine/provider/pe/sge/SgePeProviderTest.java
#	src/test/java/com/epam/grid/engine/provider/queue/sge/SgeQueueProviderTest.java
#	src/test/java/com/epam/grid/engine/service/JobOperationProviderServiceTest.java

ekazachkova commented on src/main/java/com/epam/grid/engine/service/JobOperationProviderService.java:
```diff
@@ -80,26 +81,30 @@ public Listing<Job> filter(final JobFilter jobFilter) {
-                        String.format("Id specified in %s for job removal is invalid!", deleteJobFilter));
+                        + "Either at least one `id` or `user` must be specified to delete jobs!", deleteJobFilter));
             }
+            deleteJobFilter.getIds().stream()
```
It looks this filter can be simplified: `.stream().filter(id -> id == null ||  id <= 0)`. Could you please check it?

AleksandrMikhin replied:
done

ekazachkova commented on src/test/java/com/epam/grid/engine/provider/job/sge/SgeJobProviderTest.java:
```diff
@@ -197,487 +152,237 @@ public class SgeJobProviderTest {
-                .stdErr(EMPTY_LIST)
-                .build();
-
+    public void shouldTrowWhenCommandResultFiltrationWithErrorCode() {
```
typo `Trow` -> `Throw`

AleksandrMikhin replied:
done

ekazachkova commented on src/test/java/com/epam/grid/engine/provider/job/sge/SgeJobProviderTest.java:
```diff
@@ -197,487 +152,237 @@ public class SgeJobProviderTest {
+                () -> sgeJobProvider.runJob(jobOptions, SOME_JOB_LOG_DIRECTORY_PATH));
     }
 
     private static Job runningJobTemplate() {
```
Let's move this utility private methods to the end of the file

AleksandrMikhin replied:
moved

AleksandrMikhin committed:
added ability to use test properties when running tests, refactored some tests, fixed the @code-tag inappropriate usage into javadoc (#17)

Co-authored-by: Aleksandr Mikhin <aleksandr_mikhin@epam.com>

Commit:
minor changes according to review comments

Commit:
minor changes into api-markup docs

Commit:
a one more minor change into api-markup docs

Commit:
changes according to review comments
